Common/Data/Auxiliary/FactorFile.cs

```diff
@@ -208,26 +211,127 @@ public bool HasSplitEventOnNextTradingDay(DateTime date, out decimal splitFactor
         {
             var filePath = LeanData.GenerateRelativeFactorFilePath(symbol);
+            File.WriteAllLines(filePath, ToCsvLines());
+        }
+
+        /// <summary>
+        /// Gets all of the splits and dividends represented by this factor file
+        /// </summary>
+        /// <param name="symbol">The symbol to ues for the dividend and split objects</param>
+        /// <param name="exchangeHours">Exchange hours used for resolving the previous trading day</param>
+        /// <returns>All splits and diviends represented by this factor file in chronological order</returns>
+        public List<BaseData> GetSplitsAndDividends(Symbol symbol, SecurityExchangeHours exchangeHours)
+        {
+            var dividendsAndSplits = new List<BaseData>();
+            var futureFactorFileRow = SortedFactorFileData.Last().Value;
+            for (var i = SortedFactorFileData.Count - 2; i >= 0 ; i--)
+            {
+                var row = SortedFactorFileData.Values[i];
+                var dividend = row.GetDividend(futureFactorFileRow, symbol, exchangeHours);
+                if (dividend.Distribution != 0m)
+                {
+                    dividendsAndSplits.Add(dividend);
+                }
+
+                var split = row.GetSplit(futureFactorFileRow, symbol, exchangeHours);
+                if (split.SplitFactor != 1m)
+                {
+                    dividendsAndSplits.Add(split);
+                }
+
+                futureFactorFileRow = row;
+            }
 
-            var csv = new StringBuilder();
+            return dividendsAndSplits.OrderBy(d => d.Time.Date).ToList();
+        }
+
+        /// <summary>
+        /// Creates a new factor file with the specified data applied.
+        /// Only <see cref="Dividend"/> and <see cref="Split"/> data types
+        /// will be used.
+        /// </summary>
+        /// <param name="data">The data to apply</param>
+        /// <param name="exchangeHours">Exchange hours used for resolving the previous trading day</param>
+        /// <returns>A new factor file that incorporates the specified dividend</returns>
+        public FactorFile Apply(List<BaseData> data, SecurityExchangeHours exchangeHours)
+        {
+            if (data.Count == 0)
+            {
+                return this;
+            }
+
+            var factorFileRows = new List<FactorFileRow>();
+            var lastEntry = SortedFactorFileData.Last().Value;
+            factorFileRows.Add(lastEntry);
+
+            var combinedData = GetSplitsAndDividends(data[0].Symbol, exchangeHours).Concat(data)
+                .OrderByDescending(d => d.Time.Date);
 
+            foreach (var datum in combinedData)
+            {
+                var dividend = datum as Dividend;
+                if (dividend != null)
+                {
+                    var nextEntry = lastEntry.Apply(dividend, exchangeHours);
+                    factorFileRows.Add(nextEntry);
```

Comment:
A possible case of dropped splits if the splits and dividends occur on the same date it seems?